src/Uno.UI/UI/Xaml/DependencyObjectStore.cs

```diff
@@ -1252,6 +1255,172 @@ private void CleanupInheritedProperties()
 			}
 		}
 
+		private DependencyProperty? GetCorrespondingInheritedProperty(DependencyProperty property)
```

Comment:
Is it really faster than a string lookup?

Reply:
I did quick benchmarking now. It's faster when early/middle in branching and non-significantly slower when very late in branching. I'll see if this can be improved 😕

Reply:
I had a bug in the code. Fixing it now and re-benchmarking

Reply:
```csharp
public class DPCorrespondingInheritedBenchmarks
{
	private DependencyObjectStore _gridStore;
	private DependencyObjectStore _iconElementStore;

	[GlobalSetup]
	public void Setup()
	{
		var grid = new Grid();
		_gridStore = ((IDependencyObjectStoreProvider)grid).Store;

		var iconElement = new Grid();
		_iconElementStore = ((IDependencyObjectStoreProvider)iconElement).Store;
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(Grid), UIElement.HighContrastAdjustmentProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Hardcoded()
	{
		return _gridStore.GetCorrespondingInheritedProperty(UIElement.HighContrastAdjustmentProperty);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(IconElement), IconElement.ForegroundProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Hardcoded()
	{
		return _iconElementStore.GetCorrespondingInheritedProperty(IconElement.ForegroundProperty);
	}
}
```

The mean of benchmarks:

- Benchmark_HighContrastAdjustmentProperty_Lookup: 32.0982 ns
- Benchmark_HighContrastAdjustmentProperty_Hardcoded: 24.5869 ns
- Benchmark_IconElement_ForegroundProperty_Lookup: 26.2229 ns
- Benchmark_IconElement_ForegroundProperty_Hardcoded: 55.4380 ns

The case of IconElement is where lots of type checks happen. It can potentially be further optimized though.

Reply:
Actually I'm incorrectly initializing `_iconElementStore` from a `Grid` instead of `IconElement` so it ended up falling back to the old path. We should detect when we know for sure that the old path isn't going to find anything and return null early. I'll adjust the logic more, but the initial benchmarks are somewhat promising.

Reply:
I hid the outdated comments. With the latest changes and the following benchmark:

```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using BenchmarkDotNet.Attributes;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace SamplesApp.Benchmarks.Suite.Windows_UI_Xaml.DPBench;

public class DPCorrespondingInheritedBenchmarks
{
	private DependencyObjectStore _gridStore;
	private DependencyObjectStore _iconElementStore;

	[GlobalSetup]
	public void Setup()
	{
		var grid = new Grid();
		_gridStore = ((IDependencyObjectStoreProvider)grid).Store;

		var iconElement = new IconElement();
		_iconElementStore = ((IDependencyObjectStoreProvider)iconElement).Store;
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(Grid), UIElement.HighContrastAdjustmentProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Hardcoded()
	{
		return _gridStore.GetCorrespondingInheritedProperty(UIElement.HighContrastAdjustmentProperty);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(IconElement), IconElement.ForegroundProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Hardcoded()
	{
		return _iconElementStore.GetCorrespondingInheritedProperty(IconElement.ForegroundProperty);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_For_Grid_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(Grid), IconElement.ForegroundProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_For_Grid_Hardcoded()
	{
		return _gridStore.GetCorrespondingInheritedProperty(IconElement.ForegroundProperty);
	}
}
```

Results on Skia WPF (Release) - all numers are "Mean":

- Benchmark_HighContrastAdjustmentProperty_Lookup: 38.6518 ns
- Benchmark_HighContrastAdjustmentProperty_Hardcoded: 20.5539 ns
- Benchmark_IconElement_ForegroundProperty_Lookup: 34.5967 ns
- Benchmark_IconElement_ForegroundProperty_Hardcoded: 1.9096 ns
- Benchmark_IconElement_ForegroundProperty_For_Grid_Lookup: 24.8969 ns
- Benchmark_IconElement_ForegroundProperty_For_Grid_Hardcoded: 66.2000

It's not ideal I think 😕 

I'll see if this can be improved

Reply:
Benchmark:

```csharp
public class DPCorrespondingInheritedBenchmarks
{
	private DependencyObjectStore _gridStore;
	private DependencyObjectStore _iconElementStore;

	[GlobalSetup]
	public void Setup()
	{
		var grid = new Grid();
		_gridStore = ((IDependencyObjectStoreProvider)grid).Store;

		var iconElement = new IconElement();
		_iconElementStore = ((IDependencyObjectStoreProvider)iconElement).Store;
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(Grid), UIElement.HighContrastAdjustmentProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_HighContrastAdjustmentProperty_Hardcoded()
	{
		return _gridStore.GetCorrespondingInheritedProperty(UIElement.HighContrastAdjustmentProperty);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(IconElement), IconElement.ForegroundProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_Hardcoded()
	{
		return _iconElementStore.GetCorrespondingInheritedProperty(IconElement.ForegroundProperty);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_For_Grid_Lookup()
	{
		return DependencyProperty.GetProperty(typeof(Grid), IconElement.ForegroundProperty.Name);
	}

	[Benchmark]
	public DependencyProperty Benchmark_IconElement_ForegroundProperty_For_Grid_Hardcoded()
	{
		return _gridStore.GetCorrespondingInheritedProperty(IconElement.ForegroundProperty);
	}
}
```

Results on Skia WPF (Release) - Numbers are "Mean,Media"

Benchmark_HighContrastAdjustmentProperty_Lookup: 30.7489 ns, 30.5461 ns
Benchmark_HighContrastAdjustmentProperty_Hardcoded: 25.0533 ns, 24.9849 ns
Benchmark_IconElement_ForegroundProperty_Lookup: 28.7144 ns, 28.5242 ns
Benchmark_IconElement_ForegroundProperty_Hardcoded: 1.9257 ns, 1.9191 ns
Benchmark_IconElement_ForegroundProperty_For_Grid_Lookup: 34.4348 ns, 34.3407 ns
Benchmark_IconElement_ForegroundProperty_For_Grid_Hardcoded: 40.0635 ns, 39.7477 ns

Reply:
It's interesting to see that not all of them see a substantial improvement. the 1.9ns is interesting, but the second line is curious.

Reply:
The 1.9ns is the fast early check for comparing current instance type with DP type and return the same DP if they match. This one can be added before the lookup logic.